### PR TITLE
added support for cmake "find_package"

### PR DESCRIPTION
**Pull request by prateek9623**
**Description**: 
Adds support for cmake find_package.

**Motivation and Context**
As mentioned in issue #7150 onnxruntime doesn't have support for CMake find_package, this PR adds that and also adds the CMake package version file. Now anyone can link onnxruntime like this:
```cmake
find_package(onnxruntime)
add_executable(test Source.cpp)
target_link_libraries(test PRIVATE onnxruntime::onnxruntime)
```
this also simplifies 3124

### CODE REVIEW

**Comment by prateek9623** [comment]
@snnn  I have a question are this compile definitions needs to be public? https://github.com/microsoft/onnxruntime/blob/acd9db7fadb8da618c77aef31a4df626448343e9/cmake/CMakeLists.txt#L1080

**Comment by prateek9623** [comment]
also is this file: [eigen_common_wrapper.h](https://github.com/microsoft/onnxruntime/blob/acd9db7fadb8da618c77aef31a4df626448343e9/include/onnxruntime/core/common/eigen_common_wrapper.h) is intended to be used by user or this private header as this header file include `onnxruntime_config.h` which is cmake generate file and is not exported. If it is public then we should export `onnxruntime_config.h`
https://github.com/microsoft/onnxruntime/blob/acd9db7fadb8da618c77aef31a4df626448343e9/include/onnxruntime/core/common/eigen_common_wrapper.h#L8

**Comment by snnn** [comment]
"onnxruntime_config.h"  is a tricky thing. The idea was borrowed from [autoconf](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Making-configure-Scripts.html), which is quite old. And it assumes that you won't do "build-once, run everywhere" such things. So the content of config.h is based on your current system's  configuration. If you won't copy the generated file to another file, then it's totally fine to install it to /usr/include .

**Comment by prateek9623** [comment]
> "onnxruntime_config.h" is a tricky thing. The idea was borrowed from [autoconf](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Making-configure-Scripts.html), which is quite old. And it assumes that you won't do "build-once, run everywhere" such things. So the content of config.h is based on your current system's configuration. If you won't copy the generated file to another file, then it's totally fine to install it to /usr/include .

**Comment by prateek9623** [comment]
ok, and what about other headers, like 

- `"core/framework/allocatormgr.h", "core/framework/data_transfer.h"` used by   `"core/framework/execution_provider.h"`
- `"core/framework/prepacked_weights_container.h"` used by `"core/framework/op_kernel.h"`
- `"core/framework/TensorSeq.h"` used by `"core/framework/data_types.h", "core/framework/ort_value.h"`
- `"core/framework/config_options.h"` used by `"core\framework\run_options.h"`
- `"core/platform/ort_mutex.h"` used by `"schema_registry.h"` 
- `"core/platform/threadpool.h"` used by `environment.h`
- `"thread_utils.h"` used by `""core/framework/session_options.h"` used by `graph_viewer.h`
- `path_string.h` used by `"core/common/path.h"` used by `graph.h`

**Comment by snnn** [comment]
If they need be in the right place, please help us move them.

File: onnxruntime/core/framework/onnxruntime_sequence_type_info.h
Comment (snnn) [inline onnxruntime/core/framework/onnxruntime_sequence_type_info.h:5]: Your other changes are all good. But with this line of change, I'm afraid we will need to fix our samples. https://github.com/microsoft/onnxruntime-inference-examples/tree/main/c_cxx . They use our C API and take an assumption that onnxruntime_c_api.h can be used individually without extra files(which is not true anymore) and folder structure. So that the samples looks as simple as possible. 

So, maybe we should put the C/C++ header files in a separated folder(no longer core/session) and add the new folder to the include path. 

What do you think?

Comment (prateek9623) [reply onnxruntime/core/framework/onnxruntime_sequence_type_info.h:5]: I think the current segregation of header files is good, we should keep it, and if we want to keep samples as is, we can add `/include/onnxruntime/core/session` install interface of `onnxruntime` cmake target.

**Comment by prateek9623** [comment]
@snnn Hey, this pr is ready for shared libs. Can you review it or ask someone else to review it.

**Comment by WilliamTambellini** [comment]
@prateek9623  does it now need a rebase ?

**Comment by ytaous** [comment]
> @snnn Hey, this pr is ready for shared libs. Can you review it or ask someone else to review it if you dont have time.

pls resolve the conflict files and rebuild, thx

**Comment by prateek9623** [comment]
@WilliamTambellini @ytaous i have resolved conflicts and rebase the branch, please go ahead with the review

**Comment by prateek9623** [comment]
@ytaous hi, there were some merge conflicts after the azure pipeline run. While resolving them, I found that one commit was not pushed, which I have also pushed and resolved the merge conflict. Sorry for the mistake. Would you please review it again?

**Comment by prateek9623** [comment]
Hi, build was failing because cmake export onnxruntimeTarget require shared lib target and make was not able to find it, to fix this I have added a commit.

**Comment by prateek9623** [comment]
@snnn @WilliamTambellini @ytaous Hi, this ahould be the last commit, please review it. thanks

File: cmake/CMakeLists.txt
Comment (snnn) [inline cmake/CMakeLists.txt:1355]: Some of the formatting changes, like this, seem unnecessary.

File: cmake/onnxruntime_unittests.cmake
Comment (snnn) [inline cmake/onnxruntime_unittests.cmake:116]: "MACOSX_BUNDLE_SHONNXRUNTIME_VERSION_STRING", what's this?

File: cmake/onnxruntime_config.h.in
Comment (snnn) [inline cmake/onnxruntime_config.h.in:22]: Good! Thank you!

**Comment by prateek9623** [comment]
@snnn Please rerun the azure pipelines

**Comment by prateek9623** [comment]
It looks like NUPHAR build is failing. I will try to resolve this, but can anybody help me with this, as I don't know much about NUPHAR.

**Comment by snnn** [comment]
> It looks like NUPHAR build is failing. I will try to resolve this, but can anybody help me with this, as I don't know much about NUPHAR.

@pranavsharma , can you please help ?

**Comment by prateek9623** [comment]
@snnn Ok I found the issue with Linux Nuphar builds, issue is with cmake target tvm it doesn't belong to any export. This is can be resolved by updating cmake for external/tvm using some patch or direct contribute in tvm but this still wouldn't export tvm lib because tvm subdirectory is set to "EXCLUDE_FROM_ALL" . I think the proper way would be to separate tvm from onnxruntime build and create a cmake wrapper to find tvm.

**Comment by snnn** [comment]
@xadupre  is doing some refactoring on the TVM part. #10055 You may discuss it with him.

**Comment by prateek9623** [comment]
@xadupre Hey, can you help me with this PR? Its build is failing for Linux NUPHAR build mainly because target tvm is not in any cmake export group. https://github.com/microsoft/onnxruntime-tvm/blob/4e2d707f2d2160a30375267269c4f5da9a810cfa/CMakeLists.txt#L339 
Like in: https://github.com/microsoft/onnxruntime/blob/a3cb59bc5ef273de009780ab1c906048d8a0e750/cmake/onnxruntime.cmake#L208
If we can add a export group for tvm then this PR will build successfully.

**Comment by snnn** [comment]
onnxruntime-tvm? I think we should deprecate the repo.

**Comment by prateek9623** [comment]
> onnxruntime-tvm? I think we should deprecate the repo.

I don't know much about tvm, so I can't comment on it. Maybe @xadupre can say something.

**Comment by xadupre** [comment]
> > onnxruntime-tvm? I think we should deprecate the repo.
> 
> I don't know much about tvm, so I can't comment on it. Maybe @xadupre can say something.

onnxruntime-tvm is used by Nuphar. It is a fork of tvm done 2 years ago. Nuphar is still used.

**Comment by prateek9623** [comment]
> > > onnxruntime-tvm? I think we should deprecate the repo.
> > 
> > 
> > I don't know much about tvm, so I can't comment on it. Maybe @xadupre can say something.
> 
> onnxruntime-tvm is used by Nuphar. It is a fork of tvm done 2 years ago. Nuphar is still used.

cant we update it?

**Comment by xadupre** [comment]
> > > > onnxruntime-tvm? I think we should deprecate the repo.
> > > 
> > > 
> > > I don't know much about tvm, so I can't comment on it. Maybe @xadupre can say something.
> > 
> > 
> > onnxruntime-tvm is used by Nuphar. It is a fork of tvm done 2 years ago. Nuphar is still used.
> 
> cant we update it? replace it with apache/tvm

It should be replaced but in two years time, tvm made many changes in its C++ API and there would be many changes to do. Most of the files in subfolder https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/core/codegen would be modified. I started something in this PR #10183. It is not finished.

**Comment by prateek9623** [comment]
@snnn @xadupre Just curious why is Nuphar EP and tvm is dependent of onnxruntime cmake target. As per my understanding, tvm is a dependency of Nuphar EP, and Nuphar EP can be a module target like CUDA, TensorRT EP.

**Comment by xadupre** [comment]
I don't know the reason. I was not actively working on onnxruntime when it was done.

**Comment by snnn** [comment]
> Just curious why target onnxruntime is dependent on Nuphar EP and tvm

It need to export a function from each EP, so that you can created an EP object. So it need to link to Nuphar EP.

**Comment by prateek9623** [comment]
@snnn if we are able to do that then this PR will build successfully. What do you think should we do it?

**Comment by jinfagang** [comment]
@prateek9623 @snnn when will this PR can merge? we definitely need such feature for CMake based projects.

**Comment by prateek9623** [comment]
@jinfagang This PR is ready but we cant merge as Nuphar build is failing due to tvm.

**Comment by jinfagang** [comment]
@prateek9623 before this nice PR merged in, is there any other way to link onnxruntime in cmake?

**Comment by prateek9623** [comment]
@jinfagang I use this cmake module files for the time being [Findonnxruntime.cmake](https://gist.github.com/prateek9623/976dc69dd9b4715e6fe77047c8d8890d)

**Comment by prateek9623** [comment]
@snnn hey as https://github.com/microsoft/onnxruntime/pull/10211 is merged can we retry building Nuphar target, I am guessing it should build now.
If it builds again can we revisit this PR?

**Comment by prateek9623** [comment]
@xadupre @snnn I am unable to fix Nuphar build just by making changes in onnxruntime cmake. I only see one solution to resolve this to make changes in https://github.com/microsoft/onnxruntime-tvm. Should I create PR there, similar to https://github.com/apache/tvm/pull/10097

**Comment by xadupre** [comment]
> @xadupre @snnn I am unable to fix Nuphar build just by making changes in onnxruntime cmake. I only see one solution to resolve this to make changes in https://github.com/microsoft/onnxruntime-tvm. Should I create PR there, similar to [apache/tvm#10097](https://github.com/apache/tvm/pull/10097)

That may be a good solution. As far as I know, https://github.com/microsoft/onnxruntime-tvm is only used by onnxruntime for EP Nuphar. On the other hand, I doubt anybody would use find_package(onnxruntime) to build on the top of EP Nuphar.

File: cmake/CMakeLists.txt
Comment (snnn) [inline cmake/CMakeLists.txt:1195]: Why do you change these things to private? Is it really needed?

Comment (snnn) [reply cmake/CMakeLists.txt:1195]: If some code link to this lib and forget to define ONNXRUNTIME_ENABLE_MEMLEAK_CHECK, it will crash.

**Comment by prateek9623** [comment]
@snnn @xadupre  I have created PR https://github.com/microsoft/onnxruntime-tvm/pull/15 to resolve Nuphar build. With that, I am able to build the Nuphar target also.

File: cmake/CMakeLists.txt
Comment (snnn) [inline cmake/CMakeLists.txt:1444]: Why do you remove EXCLUDE_FROM_ALL?

File: cmake/external/tvm.cmake
Comment (snnn) [inline cmake/external/tvm.cmake:32]: Is this change needed?

**Comment by gedoensmax** [comment]
I feel like this should also contain a sample/usage in https://github.com/microsoft/onnxruntime-inference-examples/tree/main/c_cxx so that the finder is actually used. 
I tried this under Windows today and it failed to find the onnxruntimeTarget.cmake:
```
  include could not find requested file:
    ./onnxruntime/build/Windows/RelWithDebInfo/RelWithDebInfo/onnxruntimeTargets.cmake
```
The file is actually located at: `onnxruntime\build\Windows\RelWithDebInfo\RelWithDebInfo\CMakeFiles\Export\ab2ff715e973603cf549f909dd4b8bcc`

**Comment by prateek9623** [comment]
@gedoensmax did you build the "install" target?

**Comment by gedoensmax** [comment]
Ok you were correct, the install target was not built. I had to do `cmake --build ./onnxruntime/build/Windows/RelWithDebInfo/RelWithDebInfo -t install`. Should this be exposed inside the python `ci_build` as a `--install` flag for example ? That did fix the `find_package` if i set the `onnxruntime_DIR` to `${CMAKE_INSTALL_PREFIX}/lib/cmake/onnxruntime/`
But it get an error that says:
```
CMake Error in cuda_provider/CMakeLists.txt:
  Imported target "onnxruntime::onnxruntime" includes non-existent path

    "C:/Users/maximilianm/repos/onnxruntime/install/include/onnxruntime/core/session"
```
which is true as the folder does not exist. 
In general it would make sense to have a sample on how to use this in the docs i think.

File: cmake/onnxruntime.cmake
Comment (gedoensmax) [inline cmake/onnxruntime.cmake:101]:
```suggestion
target_include_directories(onnxruntime PRIVATE ${ONNXRUNTIME_ROOT} PUBLIC "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/onnxruntime;${CMAKE_INSTALL_INCLUDEDIR}/onnxruntime/core>")
```
Contents of session are installed into `${CMAKE_INSTALL_INCLUDEDIR}/onnxruntime`.

**Comment by gedoensmax** [comment]
Works lika a shine for me ! This should be in master.

**Comment by princessakokosowa** [comment]
@prateek9623 Say that's I've got a project where ONNX Runtime is an external dependency and that I've cloned it to _<root_dir>/lib/_. How, according to what has been done in this thread, should I add this library to my _CMakeLists.txt_ so that it all works (**I am using your branch**)?

**Comment by prateek9623** [comment]
@princessakokosowa I am currently using this [cmake](https://github.com/microsoft/onnxruntime/files/9739266/cmake.zip) files for linking onnxruntime in my projects. This cmake module can integrate NuGet/zip release packages like in [Releases](https://github.com/microsoft/onnxruntime/releases). You can also use this. Just include this files in your CMakelist.txt and add `find_package(onnxruntime 1.8.0 COMPONENTS onnxruntime_providers_cuda OPTIONAL_COMPONENTS onnxruntime_providers_tensorrt)` statement

**Comment by princessakokosowa** [comment]
@prateek9623 thanks for the answer! Do the cmake files require changes made in the scope of this PR? Also, where am I supposed to paste the cmake files? Either in _onnxruntime_ or _onnxruntime/cmake_ directory? I'd be glad if you could also help me with that.

**Comment by prateek9623** [comment]
@princessakokosowa, you don't need this pr for those files, and the release by Microsoft should work with these files. Those files are required to be part of your project. You can refer to this [link](https://stackoverflow.com/q/10765885/6946021) for more info on using the custom "Find" module.

**Comment by keleftheriou** [comment]
Looks like the CI pipelines haven't run on the most recent changes -- is this because conflicts need to be resolved?

**Comment by fs-eire** [comment]
The conflicts should be resolved

**Comment by yeahdongcn** [comment]
When could we get this PR reviewed and merged?

**Comment by prateek9623** [comment]
@snnn I have rebased the branch, it should now create proper cmake files to support cmake "find_package". I have tested this on Linux and Windows, with CUDA and TensorRT providers.

**Comment by Timen** [comment]
Any idea when this PR will be merged?

**Comment by tianleiwu** [comment]
@prateek9623, can you merge latest main and resolve conflicts? After that, I could help start build pipelines.

**Comment by prateek9623** [comment]
@tianleiwu I have merged the latest main branch. 
I am currently trying Windows tensorrt build, and facing issues with unit test.

**Comment by mszhanyi** [comment]
azp run Zip-Nuget-Java-Nodejs Packaging Pipeline

**Comment by tianleiwu** [comment]
> I don't have enough React Native knowledge to fix these 2 remaining issues. Can anyone help with this?
It is fine since other jobs have same failure in the React pipeline. I think it is not related to this change.

**Comment by tianleiwu** [comment]
@snnn, are there other pipelines need to verify for this PR? 
I think some release pipelines (like packaging pipeline) cannot run in this PR since those require code in main/release branch.

**Comment by YUNQIUGUO** [comment]
> I don't have enough React Native knowledge to fix these 2 remaining issues. Can anyone help with this?

Yeah, the react native pipeline is currently failing due to some other checked-in changes. We are working on a fix for that.

**Review by tianleiwu — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Comment by QiuYilin** [comment]
It seems that the target onnxruntime does not contain providers, which causes onnxruntime_providers.dll and onnxruntime_provider_cuda.dll to be lost when copying the target to the root directory.

**Comment by prateek9623** [comment]
@QiuYilin I did that because other providers' targets are loaded at the runtime. But those targets should be installed if the "install" target is built.

**Comment by Artalus** [comment]
@prateek9623 can you or someone else please explain, why the header structure had to be changed?..

In `1.15.1` and before, no matter if building ORT as shared or static, we always had `include/onnxruntime/core/session/onnxruntime_cxx_api.h` for example, same as it was in the ORT repo source tree.

But now, since in `cmake/onnxruntime_session.cmake` you have moved `install(DIRECTORY ${PROJECT_SOURCE_DIR}/../include/onnxruntime/core/session ...` under `if (NOT onnxruntime_BUILD_SHARED_LIB)`, the aforementioned header may move around if ORT was built as shared. So in `1.16.3` it may become `include/onnxruntime/onnxruntime_cxx_api.h` instead.

Do we now have to use something like this in our code:
```c++
#ifdef ORT_SHARED
#  include <onnxruntime/onnxruntime_cxx_api.h>
#else
#  include <onnxruntime/core/session/onnxruntime_cxx_api.h>
#endif
```
, or am I missing something?

**Comment by prateek9623** [comment]
@Artalus With this PR I tried to replicate prebuilt shared lib releases. And have not disturbed static build.
@snnn do you have any comments on this?

**Comment by thiagocrepaldi** [comment]
@snn could the cmake files be added to the tar.gz file so that people can download ort and link to it without having to clone the repo?

**Comment by WilliamTambellini** [comment]
I agree with @thiagocrepaldi

**Pull request merged**
